test/core/transport/chaotic_good/end2end/chaotic_good_server_test.cc

```diff
@@ -0,0 +1,170 @@
+#include "src/core/lib/resource_quota/arena.h"
+#include "src/core/lib/resource_quota/memory_quota.h"
+#include "src/core/lib/resource_quota/resource_quota.h"
+#include "src/core/lib/slice/slice.h"
+#include "src/core/lib/slice/slice_buffer.h"
+#include "src/core/lib/surface/server.h"
+#include "src/core/lib/transport/metadata_batch.h"
+#include "src/core/lib/transport/promise_endpoint.h"
+#include "src/core/lib/transport/transport.h"
+#include "src/core/lib/uri/uri_parser.h"
+#include "src/cpp/server/secure_server_credentials.h"
+#include "test/core/util/port.h"
+
+namespace grpc_core {
+namespace chaotic_good {
+namespace testing {
+using grpc_event_engine::experimental::EventEngine;
+class ChaoticGoodServerTest : public ::testing::Test {
+ public:
+  ChaoticGoodServerTest() {
+    event_engine_ = std::shared_ptr<EventEngine>(
+        grpc_event_engine::experimental::CreateEventEngine());
+    channel_args_ = channel_args_.SetObject(event_engine_);
+    resource_quota_ = ResourceQuota::Default();
+    channel_args_ = channel_args_.Set(GRPC_ARG_RESOURCE_QUOTA, resource_quota_);
+    StartServer();
+    ConstructConnector();
+  }
+  ~ChaoticGoodServerTest() override { core_server_->StopListening(); }
+
+  void StartServer() {
+    port_ = grpc_pick_unused_port_or_die();
+    addr_ = absl::StrCat("ipv6:[::1]:", port_);
+    ExecCtx exec_ctx;
+    server_ = grpc_server_create(nullptr, nullptr);
+    core_server_ = Server::FromC(server_);
+    listener_ = std::make_shared<ChaoticGoodServerListener>(core_server_,
+                                                            channel_args_);
+    auto port = listener_->Bind(addr_.c_str());
+    EXPECT_TRUE(port.ok());
+    EXPECT_EQ(port.value(), port_);
+    auto status = listener_->StartListening();
+    EXPECT_TRUE(status.ok());
+  }
+
+  void ConstructConnector() {
+    auto uri = URI::Parse(addr_);
+    GPR_ASSERT(uri.ok());
+    GPR_ASSERT(grpc_parse_uri(*uri, &resolved_addr_));
+    args_.address = &resolved_addr_;
+    args_.deadline = Timestamp::Now() + Duration::Seconds(5);
+    args_.channel_args = channel_args_;
+    connector_ = std::make_shared<ChaoticGoodConnector>();
+  }
+
+ protected:
+  static void OnConnectingFinished(void* arg, grpc_error_handle error) {
+    ChaoticGoodServerTest* self = static_cast<ChaoticGoodServerTest*>(arg);
+    self->connect_finished_.Notify();
+  }
+  grpc_server* server_;
+  Server* core_server_;
+  ChannelArgs channel_args_;
+  ChaoticGoodConnector::Args args_;
+  ChaoticGoodConnector::Result connecting_result_;
+  grpc_closure on_connecting_finished_;
+  int port_;
+  std::string addr_;
+  grpc_resolved_address resolved_addr_;
+  std::shared_ptr<ChaoticGoodServerListener> listener_;
+  std::shared_ptr<ChaoticGoodConnector> connector_;
+  std::shared_ptr<EventEngine> event_engine_;
+  std::unique_ptr<MemoryQuota> memory_quota_;
+  ResourceQuotaRefPtr resource_quota_;
+  Notification connect_finished_;
+};
+
+TEST_F(ChaoticGoodServerTest, Connect) {
```

Comment:
Can you add a test where the client connects to an address with no server listening on it. In that case it should fail gracefully

Reply:
+1.

Also a test of two clients connecting simultaneously.

And a test of the server shutting down after the first connection is successfully created.